### PR TITLE
feat(pluginpreset): add clusterOptionOverrides

### LINKED ISSUE #491: [FEAT] - PluginPreset Overrides
### Priority

(Medium) I'm annoyed but I'll live

### Description

PluginPresets are a way to define a default configuration for a Plugin, which is deployed to all clusters matching the PluginPreset's selector.
The issue now is that there are some cases where the default configuration between two clusters only differs in one or very few values, e.g. a cluster-specific secret. This is currently not possible with PluginPresets, as they are applied with the same configuration to all clusters matching the selector.

With [ADR 007](https://github.com/cloudoperators/documentation/blob/main/architecture-decision-records/007-greenhouse-plugin-overrides.md#additonal-clusteroptionoverrides-field-on-pluginpreset) a decision was made how to overcome this shortcoming.
It describes adding an additional field called `ClusterOptionOverrides`. This field contains a list of overrides that override the values of the PluginSpec for the corresponding cluster.

```golang
type ClusterOptionOverride struct {
  ClusterName string `json:"clusterName"`
  Overrides []PluginOptionValue `json:"overrides"`
}
```

## Acceptance Criteria

- [ ] It is possible to override values of the PluginSpec from a PluginPreset
- [ ] Documentation of the new Feature
- [ ] Integration Tests showcasing that Adding, Editing and Removing Overrides works.

### PR DESCRIPTION
**Pull request by gciezkowski-acc**
## Description
I added the field clusterOptionOverrides to Plugin Preset Spec struct.

## What type of PR is this? (check all applicable)

- [X] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related Issue # (issue)
- Closes #491 
- Fixes # (issue)

> Remove if not applicable

## Added tests?

- [X] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added to documentation?

- [ ] 📜 README.md
- [X] 🤝 Documentation pages updated
- [ ] 🙅 no documentation needed
- [ ] (if applicable) generated OpenAPI docs for CRD changes

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: pkg/controllers/plugin/pluginpreset_controller_test.go [hunk 1]
```diff
@@ -242,6 +242,151 @@ var _ = Describe("PluginPreset Controller Lifecycle", Ordered, func() {
...
+					OptionValues: []greenhousev1alpha1.PluginOptionValue{
+						{
+							Name:  "option-1",
+							Value: asAPIextensionJSON(2),
+						},
+					},
+				},
+			},
+		),
+		Entry("with defined pluginPresetOverrides but for another cluster",
```
Comment (IvoGoman) [inline pkg/controllers/plugin/pluginpreset_controller_test.go:314]:
```suggestion
		Entry("with defined pluginPresetOverrides for the correct cluster",
```

File: pkg/controllers/plugin/pluginpreset_controller_test.go [hunk 2]
```diff
@@ -242,6 +242,151 @@ var _ = Describe("PluginPreset Controller Lifecycle", Ordered, func() {
...
+					OptionValues: []greenhousev1alpha1.PluginOptionValue{
+						{
+							Name:  "option-1",
+							Value: asAPIextensionJSON(1),
+						},
+					},
+				},
+			},
+		),
+		Entry("with defined pluginPresetOverrides but for another cluster",
```
Comment (IvoGoman) [inline pkg/controllers/plugin/pluginpreset_controller_test.go:353]:
```suggestion
		Entry("with defined pluginPresetOverrides for the cluster",
```

File: pkg/controllers/plugin/pluginpreset_controller.go [hunk 3]
```diff
@@ -218,6 +221,30 @@ func (r *PluginPresetReconciler) reconcilePluginPreset(ctx context.Context, pres
...
+
+	// when plugin is running on different cluster then defined in
+	if index == -1 {
+		return
+	}
+
+	// overrides value
+	for _, overrideValue := range preset.Spec.ClusterOptionOverrides[index].Overrides {
+		valueIndex := slices.IndexFunc(plugin.Spec.OptionValues, func(value greenhousev1alpha1.PluginOptionValue) bool {
+			return true
```
[the remark this answers is not in the document]
Comment (gciezkowski-acc) [reply pkg/controllers/plugin/pluginpreset_controller.go]: Done

File: pkg/controllers/plugin/pluginpreset_controller_test.go (hunk 1 above)
Comment (gciezkowski-acc) [reply pkg/controllers/plugin/pluginpreset_controller_test.go]: Done

File: pkg/controllers/plugin/pluginpreset_controller_test.go (hunk 2 above)
Comment (gciezkowski-acc) [reply pkg/controllers/plugin/pluginpreset_controller_test.go]: Done

File: pkg/apis/greenhouse/v1alpha1/pluginpreset_types.go [hunk 4]
```diff
@@ -20,6 +20,15 @@ type PluginPresetSpec struct {
 
 	// ClusterSelector is a label selector to select the clusters the plugin bundle should be deployed to.
 	ClusterSelector metav1.LabelSelector `json:"clusterSelector"`
+
+	// ClusterOptionOverrides define plugin option values to override by the PluginPreset
+	ClusterOptionOverrides []ClusterOptionOverride `json:"clusterOptionOverrides"`
+}
+
+// ClusterOptionOverride defines which plugin option should be override in which cluster
+type ClusterOptionOverride struct {
```
Comment (IvoGoman) [inline pkg/apis/greenhouse/v1alpha1/pluginpreset_types.go:29]: This field should be optional
```suggestion
// +kubebuilder:validation:Optional
type ClusterOptionOverride struct {
```

**Review by IvoGoman — CHANGES_REQUESTED** [review]
Could you run `make generate` to update also the generated CRDs?

File: pkg/admission/pluginpreset_webhook_test.go [hunk 5]
```diff
@@ -58,7 +58,8 @@ var _ = Describe("PluginPreset Admission Tests", Ordered, func() {
 				Namespace: test.TestNamespace,
 			},
 			Spec: greenhousev1alpha1.PluginPresetSpec{
-				ClusterSelector: metav1.LabelSelector{MatchLabels: map[string]string{"foo": "bar"}},
+				ClusterSelector:        metav1.LabelSelector{MatchLabels: map[string]string{"foo": "bar"}},
+				ClusterOptionOverrides: []greenhousev1alpha1.ClusterOptionOverride{},
```
Comment (IvoGoman) [inline pkg/admission/pluginpreset_webhook_test.go:62]: This won't be necessary after making the field optional. The overrides should not be required

File: pkg/apis/greenhouse/v1alpha1/pluginpreset_types.go (hunk 4 above)
Comment (gciezkowski-acc) [reply pkg/apis/greenhouse/v1alpha1/pluginpreset_types.go:31]: Done

**Comment by IvoGoman** [comment]
> Could you run `make generate` to update also the generated CRDs?

This still needs to be done ^^

**Review by IvoGoman — APPROVED** [review]
Thanks, looks good.

**Pull request merged**